src/metrics.jl

```diff
@@ -100,216 +182,156 @@ struct MeanSqDeviation <: SemiMetric end
-see [`Euclidean`](@ref).
-"""
-SqEuclidean() = SqEuclidean(0)
-
-"""
-	    PeriodicEuclidean(L)
-
-Create a Euclidean metric on a rectangular periodic domain (i.e., a torus or
-a cylinder). Periods per dimension are contained in the vector `L`:
-```math
-\\sqrt{\\sum_i(\\min\\mod(|x_i - y_i|, p), p - \\mod(|x_i - y_i|, p))^2}.
-```
-For dimensions without periodicity put `Inf` in the respective component.
-
-# Example
-```jldoctest
-julia> x, y, L = [0.0, 0.0], [0.75, 0.0], [0.5, Inf];
-
-julia> evaluate(PeriodicEuclidean(L), x, y)
-0.25
-```
-"""
-PeriodicEuclidean() = PeriodicEuclidean(Int[])
+const weightedmetrics = (WeightedEuclidean,WeightedSqEuclidean,WeightedCityblock,WeightedMinkowski,WeightedHamming)
+const UnionWeightedMetrics = Union{weightedmetrics...}
+const UnionMetrics = Union{metrics...,weightedmetrics...}
 
 ###########################################################
 #
 #  Implementations
 #
 ###########################################################
 
-const ArraySlice{T} = SubArray{T,1,Array{T,2},Tuple{Base.Slice{Base.OneTo{Int}},Int},true}
+parameters(::UnionMetrics) = nothing
+parameters(d::PeriodicEuclidean) = d.periods
+parameters(d::UnionWeightedMetrics) = d.weights
+
+result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray) =
+    result_type(dist, a, b, parameters(dist))
+result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray, ::Nothing) =
+    typeof(_evaluate(dist, oneunit(eltype(a)), oneunit(eltype(b))))
+result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray, p) =
+    typeof(_evaluate(dist, oneunit(eltype(a)), oneunit(eltype(b)), oneunit(eltype(p))))
 
-@inline parameters(::UnionMetrics) = nothing
+Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a::AbstractArray, b::AbstractArray)
+    _evaluate(d, a, b, parameters(d))
+end
+_evaluate(dist::UnionMetrics, a::Number, b::Number) = _evaluate(dist, a, b, parameters(dist))
 
 # breaks the implementation into eval_start, eval_op, eval_reduce and eval_end
 
-# Specialized for Arrays and avoids a branch on the size
-@inline Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a::Union{Array, ArraySlice}, b::Union{Array, ArraySlice})
+Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a::AbstractArray, b::AbstractArray, ::Nothing)
     @boundscheck if length(a) != length(b)
         throw(DimensionMismatch("first array has length $(length(a)) which does not match the length of the second, $(length(b))."))
     end
-    p = parameters(d)
-    @boundscheck if p !== nothing
-        length(a) != length(p) && throw(DimensionMismatch("arrays have length $(length(a)) but parameters have length $(length(p))."))
-    end
     if length(a) == 0
         return zero(result_type(d, a, b))
     end
     @inbounds begin
         s = eval_start(d, a, b)
-        if p === nothing
+        if IndexStyle(a, b) === IndexLinear() || size(a) == size(b)
             @simd for I in 1:length(a)
```

Comment:
This is more efficient for `IndexCartesian`:
```suggestion
            @simd for I in eachindex(a, b)
```

Reply:
Indeed, in the little benchmark it's performing faster. 👍

Reply:
Wow, that is actually very interesting, because the reason `1:length(a)` (to index into _both_ `a` _and_ `b` was added in the first place was https://github.com/JuliaLang/julia/issues/28226. However, now `eachindex(a, b)` seems to work nicely:
```julia
julia> @btime colwise(dd, Xcol,Ycol);
  923.614 μs (2 allocations: 781.33 KiB)

julia> @btime colwise_naive(Xcol,Ycol);
  975.183 μs (2 allocations: 781.33 KiB)
```